docs/source/posts/2024/2024-06-06-week-1-robin.rst

```diff
@@ -0,0 +1,106 @@
+Performance wise, it ranks 11th on the `MTEB leaderboard <https://huggingface.co/spaces/mteb/leaderboard>`_. It is a very interesting model due to its size:performance ratio.
+
+4) **Hosting the embedding model**
+
+Hosting this sentence-transformer model was confusing. For some reason, the HF spaces were blocking the Python script from writing on ``.cache`` folder. Later I figured out the reason and fixed it.
+I've hosted 5 gunicorn workers to serve 5 parallel requests at a time. Since the model is small, this is possible.
+
+5) **Hosting the database endpoint**
+
+I wrapped the pinecone db api into an endpoint so it'll be easy to query and receive the results.
+It is also configured to accept 5 concurrent requests although I could increase it a lot more.
+
+I upserted docstrings from ``fury/actor.py`` into the vector db for testing. So now, whenever you ask a question the model will use some ``actor.py`` function to give you an answer. For now, it could be used like a semantic function search engine.
+
+I decided to abstract the db endpoint to reduce the dependency on one provider. We can swap the providers as required and keep all other features running.
+
+6) **Hosting Discord Bot**
+
+So with this, all the endpoints are finally online. The bot has some issues, it is going offline midway for some reason. I'll have to see why that happens.
+
+For some reason, huggingface spaces decided to not start the bot script. Later a community admin from huggingface told me to use their official bot implementation as a reference. This is why I had to use threading and gradio to get the bot running (migrating to docker can be done, but this is how they did it and I just took that for now).
+
+Huggingface spaces need a script to satisfy certain criteria to allow them to run, one of them is a non-blocking I/O on the main loop. So I had to move the discord bot to a new thread.
+
+Connecting all of them together!
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+So now we have 4 hosted services:
+ - Discord Bot
+ - LLM API
+ - Embeddings API
+ - Database API
+
+Now we'll have to connect them all to get an answer to the user query.
+
+This is the current architecture, there's a lot of room for improvement here.
+
+
+   .. raw:: html
+
+      <iframe src="https://drive.google.com/file/d/1HIrIJcVNZAyZGHefistB8qrS5maeE7iG/preview" width="640" height="480" allow="autoplay"></iframe>
+
+The Language Model takes the context and the user query, combines them to form an answer and returns to the user through discord (for now). Maybe moving the core logic from discord bot to a separate node might be good, and connect discord/github/X to that node.
+The database takes embeddings and do an Approximate Nearest Neighbor search (a variant of KNN) and returns top-k results (k=3 for now).
+
+  .. raw:: html
+
+   <iframe src="https://drive.google.com/file/d/1k8nVQ_7-Jl_1P78PGuhjUxu5rJEc8TZn/preview" width="640" height="390" frameborder="0" allowfullscreen></iframe>
```

Comment:
do not use a drive link. Please, upload the figure in https://github.com/fury-gl/fury-communication-assets
and then use the link from github. 

Another alternative is to host the figure here: https://github.com/fury-gl/fury/tree/master/docs/source/_static/images

Reply:
@skoudoro, I was unable to embed the video (mp4), therefore I embedded the video like this `https://github.com/robinroy03/fury-discord-bot/assets/115863770/48f1136d-18a5-45ee-aa22-0a3f6426d575`, GitHub generates this when we add videos directly to a file.

I did not find other methods.

Reply:
the real link seems to be https://github-production-user-asset-6210df.s3.amazonaws.com/115863770/337997458-48f1136d-18a5-45ee-aa22-0a3f6426d575.mp4?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA/20240610/us-east-1/s3/aws4_request&X-Amz-Date=20240610T083259Z&X-Amz-Expires=300&X-Amz-Signature=16c7c80270d05413508fd4b980002e0374222eab82669551c184f51a80cb1500&X-Amz-SignedHeaders=host&actor_id=23106443&key_id=0&repo_id=805183067